thirdparty/faiss/faiss/IndexFlat.cpp

```diff
@@ -36,7 +38,11 @@ void IndexFlat::search(
 
     if (metric_type == METRIC_INNER_PRODUCT) {
         float_minheap_array_t res = {size_t(n), size_t(k), labels, distances};
-        knn_inner_product(x, get_xb(), d, n, ntotal, &res, bitset);
+        if (is_cosine) {
+            knn_cosine(x, get_xb(), d, n, ntotal, &res, bitset);
```

Comment:
knn_cosine is a slow function which is a compromise for brute force, and will compute the norm each time distance is calculated. It's better to pre-compute norms of all base vectors and use them

Reply:
I know that, this is a middle-solution, I'll roll back some un-necessary faiss code change, then implement the flat with norm data solution